File: backend/internal/pricing/promotion/entity/ownership_status.go
package entity

import (
	"fmt"
)

// OwnershipStatus represents the status of a promotion ownership (entitlement).
// Lifecycle: available -> consumed/expired/cancelled
type OwnershipStatus string

const (
	// OwnershipStatusAvailable means the ownership is ready to use.
	// User can activate promotions with this ownership.
	OwnershipStatusAvailable OwnershipStatus = "available"

	// OwnershipStatusConsumed means the ownership's duration has been fully consumed.
	// Terminal state.
	OwnershipStatusConsumed OwnershipStatus = "consumed"

	// OwnershipStatusExpired means the ownership's validity window has passed.
	// Terminal state.
	OwnershipStatusExpired OwnershipStatus = "expired"

	// OwnershipStatusCancelled means the ownership was cancelled by admin or user.
	// Terminal state.
	OwnershipStatusCancelled OwnershipStatus = "cancelled"
)

// IsValid returns true if the ownership status is valid.
func (s OwnershipStatus) IsValid() bool {
	switch s {
	case OwnershipStatusAvailable, OwnershipStatusConsumed, OwnershipStatusExpired, OwnershipStatusCancelled:
		return true
	default:
		return false
	}
}

// IsTerminal returns true if the status is a terminal state (no further transitions possible).
func (s OwnershipStatus) IsTerminal() bool {
	return s == OwnershipStatusConsumed || s == OwnershipStatusExpired || s == OwnershipStatusCancelled
}

// CanActivate returns true if the ownership can be used to activate a promotion.
func (s OwnershipStatus) CanActivate() bool {
	return s == OwnershipStatusAvailable
}

// String returns the string representation of the ownership status.
func (s OwnershipStatus) String() string {
	return string(s)
}

// InvalidOwnershipStatusError is returned when an invalid ownership status is provided.
type InvalidOwnershipStatusError struct {
	Status OwnershipStatus
}

func (e *InvalidOwnershipStatusError) Error() string {
	return fmt.Sprintf("invalid ownership status: %s", e.Status)
}

// OwnershipTransitionError is returned when attempting an invalid status transition.
type OwnershipTransitionError struct {
	From OwnershipStatus
	To   OwnershipStatus
}

func (e *OwnershipTransitionError) Error() string {
	return fmt.Sprintf("invalid ownership status transition: %s -> %s", e.From, e.To)
}

File: apps/mobile/lib/domains/commerce/pricing/promotion/domain/entities/ownership_status.dart
/// Ownership status enum.
///
/// Lifecycle: available -> consumed/expired/cancelled
enum OwnershipStatus {
  /// Ownership is ready to use
  available('available'),

  /// Duration has been fully consumed
  consumed('consumed'),

  /// Validity window has expired
  expired('expired'),

  /// Ownership was cancelled
  cancelled('cancelled');

  const OwnershipStatus(this.value);

  /// The API value for this enum
  final String value;

  /// Parses ownership status from string
  static OwnershipStatus fromString(String value) {
    return OwnershipStatus.values.firstWhere(
      (status) => status.value == value,
      orElse: () => OwnershipStatus.available,
    );
  }
}

/// Extension for OwnershipStatus utility methods
extension OwnershipStatusX on OwnershipStatus {
  /// Whether this is a terminal state (no further transitions possible)
  bool get isTerminal =>
      this == OwnershipStatus.consumed ||
      this == OwnershipStatus.expired ||
      this == OwnershipStatus.cancelled;

  /// Whether this ownership can be used to activate a promotion
  bool get canActivate => this == OwnershipStatus.available;
}
